File: data/derived_cot/rq1_traces/translation/glm5.1/py_to_cpp/code_output/MetricsCalculator2.cpp
#include <vector>
#include <utility>
#include <numeric>

class MetricsCalculator2 {
public:
    using DataPoint = std::pair<std::vector<int>, int>;
    using ResultType = std::pair<double, std::vector<double>>;

    MetricsCalculator2() = default;

    static ResultType mrr(const DataPoint& data) {
        const auto& [sub_list, total_num] = data;
        if (total_num == 0) {
            return {0.0, {0.0}};
        }
        double mr = 0.0;
        for (size_t i = 0; i < sub_list.size(); ++i) {
            double val = static_cast<double>(sub_list[i]) / static_cast<double>(i + 1);
            if (val > 0.0) {
                mr = val;
                break;
            }
        }
        return {mr, {mr}};
    }

    static ResultType mrr(const std::vector<DataPoint>& data) {
        if (data.empty()) {
            return {0.0, {0.0}};
        }
        std::vector<double> separate_result;
        for (const auto& [sub_list, total_num] : data) {
            double mr = 0.0;
            if (total_num != 0) {
                for (size_t i = 0; i < sub_list.size(); ++i) {
                    double val = static_cast<double>(sub_list[i]) / static_cast<double>(i + 1);
                    if (val > 0.0) {
                        mr = val;
                        break;
                    }
                }
            }
            separate_result.push_back(mr);
        }
        double mean = std::accumulate(separate_result.begin(), separate_result.end(), 0.0)
                      / static_cast<double>(separate_result.size());
        return {mean, separate_result};
    }

    static ResultType map(const DataPoint& data) {
        const auto& [sub_list, total_num] = data;
        if (total_num == 0) {
            return {0.0, {0.0}};
        }
        std::vector<double> right_ranking_list;
        right_ranking_list.reserve(sub_list.size());
        int count = 1;
        for (int t : sub_list) {
            if (t == 0) {
                right_ranking_list.push_back(0.0);
            } else {
                right_ranking_list.push_back(static_cast<double>(count));
                ++count;
            }
        }
        double ap = 0.0;
        for (size_t i = 0; i < right_ranking_list.size(); ++i) {
            ap += right_ranking_list[i] / static_cast<double>(i + 1);
        }
        ap /= static_cast<double>(total_num);
        return {ap, {ap}};
    }

    static ResultType map(const std::vector<DataPoint>& data) {
        if (data.empty()) {
            return {0.0, {0.0}};
        }
        std::vector<double> separate_result;
        for (const auto& [sub_list, total_num] : data) {
            double ap = 0.0;
            if (total_num != 0) {
                std::vector<double> right_ranking_list;
                right_ranking_list.reserve(sub_list.size());
                int count = 1;
                for (int t : sub_list) {
                    if (t == 0) {
                        right_ranking_list.push_back(0.0);
                    } else {
                        right_ranking_list.push_back(static_cast<double>(count));
                        ++count;
                    }
                }
                for (size_t i = 0; i < right_ranking_list.size(); ++i) {
                    ap += right_ranking_list[i] / static_cast<double>(i + 1);
                }
                ap /= static_cast<double>(total_num);
            }
            separate_result.push_back(ap);
        }
        double mean = std::accumulate(separate_result.begin(), separate_result.end(), 0.0)
                      / static_cast<double>(separate_result.size());
        return {mean, separate_result};
    }
};

File: data/derived_cot/rq3_prompting/results/translation/qwen/cpp_to_py/code_output/MetricsCalculator2.py
class MetricsCalculator2:
    @staticmethod
    def mrr(data):
        if not data:
            return (0.0, [0.0])
        
        sum_mrr = 0.0
        individual_mrr = []
        
        for vec, k in data:
            if k <= 0 or not vec:
                individual_mrr.append(0.0)
            else:
                try:
                    pos = vec.index(1)
                    mrr_val = 1.0 / (pos + 1)
                    individual_mrr.append(mrr_val)
                    sum_mrr += mrr_val
                except ValueError:
                    individual_mrr.append(0.0)
        
        average_mrr = sum_mrr / len(data) if data else 0.0
        return (average_mrr, individual_mrr)
    
    @staticmethod
    def map(data):
        if not data:
            return (0.0, [0.0])
        
        separate_result = []
        
        for sub_list, total_num in data:
            if total_num == 0:
                separate_result.append(0.0)
                continue
            
            n = len(sub_list)
            ranking_array = [1.0 / (i + 1) for i in range(n)]
            right_ranking_list = [0.0] * n
            count = 1
            
            for i in range(n):
                if sub_list[i] != 0:
                    right_ranking_list[i] = count
                    count += 1
            
            ap = sum(x * y for x, y in zip(right_ranking_list, ranking_array)) / total_num
            separate_result.append(ap)
        
        mean_ap = sum(separate_result) / len(separate_result) if separate_result else 0.0
        return (mean_ap, separate_result)